demos/common/utilities/sample_adu_jws_mbedtls.c

```diff
@@ -0,0 +1,796 @@
+    int32_t outJWSPayloadLength;
+    int32_t outJWSSignatureLength;
+    int32_t outJWKHeaderLength;
+    int32_t outJWKPayloadLength;
+    int32_t outJWKSignatureLength;
+    az_span kidSpan = AZ_SPAN_EMPTY;
+    az_span sha256Span = AZ_SPAN_EMPTY;
+    az_span xBase64EncodedNSpan = AZ_SPAN_EMPTY;
+    az_span xBase64EncodedESpan = AZ_SPAN_EMPTY;
+    az_span xAlgSpan = AZ_SPAN_EMPTY;
+    az_span xJWKManifestSpan = AZ_SPAN_EMPTY;
+
+    if( ulScratchBufferLength < jwsSHA_CALCULATION_SCRATCH_SIZE )
+    {
+        LogError( ( "[JWS] Scratch buffer size too small" ) );
+        return eAzureIoTErrorFailed;
+    }
+
+    uint8_t * ucPersistentScratchSpaceHead = pucScratchBuffer;
+    uint8_t * ucReusableScratchSpaceRoot = ucPersistentScratchSpaceHead + jwsJWS_HEADER_SIZE + jwsJWK_PAYLOAD_SIZE;
+    uint8_t * ucReusableScratchSpaceHead = ucReusableScratchSpaceRoot;
+
+    /* Partition the scratch buffer */
+
+    /*------------------- Parse and Decode the JWS Header ------------------------*/
+
+    xResult = prvSplitJWS( pucJWS, ulJWSLength,
+                           &pucBase64EncodedHeader, &ulBase64EncodedHeaderLength,
+                           &pucBase64EncodedPayload, &ulBase64EncodedPayloadLength,
+                           &pucBase64EncodedSignature, &ulBase64SignatureLength );
+
+    if( xResult != eAzureIoTSuccess )
+    {
+        LogError( ( "[JWS] prvSplitJWS failed: result %i", xResult ) );
+        return xResult;
+    }
+
+    prvSwapToUrlEncodingChars( pucBase64EncodedSignature, ulBase64SignatureLength );
+
+    /* Note that we do not use mbedTLS to base64 decode values since we need the ability to assume padding characters. */
+    /* mbedTLS will stop the decoding short and we would then need to add in the remaining characters. */
+    ucJWSHeader = ucPersistentScratchSpaceHead;
+    ucPersistentScratchSpaceHead += jwsJWS_HEADER_SIZE;
+    xCoreResult = az_base64_decode( az_span_create( ucJWSHeader, jwsJWS_HEADER_SIZE ),
+                                    az_span_create( pucBase64EncodedHeader, ulBase64EncodedHeaderLength ),
+                                    &outJWSHeaderLength );
+
+    if( az_result_failed( xCoreResult ) )
+    {
+        LogError( ( "[JWS] az_base64_decode failed: result %i", xCoreResult ) );
+
+        if( xCoreResult == AZ_ERROR_NOT_ENOUGH_SPACE )
+        {
+            LogError( ( "[JWS] Decode buffer was too small: %i bytes", jwsJWS_HEADER_SIZE ) );
+        }
+
+        return eAzureIoTErrorFailed;
+    }
+
+    /*------------------- Parse JSK JSON Payload ------------------------*/
+
+    /* The "sjwk" is the signed signing public key */
+    AzureIoTJSONReader_Init( &xJSONReader, ucJWSHeader, outJWSHeaderLength );
+
+    if( prvFindJWSValue( &xJSONReader, &xJWKManifestSpan ) != 0 )
+    {
+        LogError( ( "Error finding sjwk value in payload" ) );
+        return eAzureIoTErrorFailed;
+    }
+
+    /*------------------- Base64 Decode the JWK Payload ------------------------*/
```

Comment:
This is splitting / parsing the payload but not Base64 decoding it on line right below.